fixed a few issues reported by golint and added the ability to do strict/non-strict filtering of IPs in order to accomodate proxied requests

fboyer opened the pull request:
Here is a quick fix for to support filtering of IPs as reported in the issue #3 . 

Please note that, by default, the middleware will now behave the same was as the logs in caddy which means that the "strict" option will be turned 'off'. (i.e.: 'X-Forwarded-For' client ip will override the req.RemoteAddr if, and only if, said header is non-empty) This might be considered a "breaking change" so let me know if this is problematic for you.

Also, I added a few tests but they are not full-blown http test as you already have other tests covering this. Since these existing tests did not break, I only added a bunch of assertions covering the new function I added to extract the proper client ip.

pyed commented:
I think your addition should be the default and the only option, can you think of a use-case that would make `strict = on` useful ?

fboyer commented:
Well I fixed it as per your suggestion. 

I would tend to agree with you as the only use-case I could think of would be to force people to go through a certain server/proxy for which you would want to strictly base yourself on the remote IP but then again I do not know if that is a real case scenario.

Also shorten my commit message which was crazy long. :)

fboyer commented:
I should have written "...force people to go through certain server/proxy without needing to put restrictions on the real client IP..."

fboyer commented:
I was wondering the same but then I looked up [Header.Get](https://golang.org/pkg/net/http/#Header.Get) and it seems to indicate that it will only return the first value. 

I also inspired myself from the following code from Caddy: https://github.com/mholt/caddy/blob/e6ba930e652ed9cfa0ba9a465e4ebbf257ad679a/middleware/replacer.go#L46-L55

As for your last comment about setting ip to ```r.RemoteAddr``` and then changing it to ```fwdFor``` if it is not empty, I am not sure that ```fwdFor``` qualifies as the default case as you seem to indicate. If I am not mistaken, this header should only be added to the request headers if the request goes through a proxy server that supports it.

I could also "optimize" this by doing an if/else in order to eliminate the "default" assignment. I must say I do not code very often in golang so I might not be aware of the "best practice".

Let me know and I'll gladly fix everything up! :)

fboyer commented:
Sorry messed up on the last rebase, I have cleaned up everything, relinted everything as per ```golint``` and removed the extra assignment as you suggested in ```getClientIP```.

Let me know if there is something else.

fboyer commented:
Moved the ```var err error``` in the else statement as you suggested.

pyed commented:
probably you missed my other code comment, you defined `getClientIP` but you haven't used it :)

fboyer commented:
And that is why I should have used full blown http tests instead of unit tests which would have caught this. :|

If everything is ok this time, you could merge this and I'll add additional test cases to your ```TestIPs``` test later on.

pyed commented:
thank you for your contribution
